rsconf/component/jupyterhub.py

```diff
@@ -207,10 +182,157 @@ def _users_for_groups(groups, t, n):
+
+
+    Will add the `host_home_d` to `guest_home_d` binding if not
+    included already in `volumes`.
+
+    Args:
+        volumes (PKDict): map of host to guest bindings with mode (see example)
+        user_groups (PKDict): map of group name to user list (see example)
+        host_home_d (py.path): root of user directories (e.g. /srv/jupyterhub/user)
+        guest_home_d (py.path): guest users's home directory (e.g. /home/vagrant)
+    """
+
+    def __init__(self, volumes, user_groups, host_home_d, guest_home_d):
+        def _cfg_to_list():
+            res = []
+            for n, v in self._cfg.items():
+                res.append(
+                    PKDict(
+                        host=n,
+                        guest=v.bind,
+                        rw=_users(v.mode.rw, n),
+                        ro=_users(v.mode.ro, n),
+                    )
+                )
+            return res
+
+        def _default_mode():
+            return PKDict(rw=[DEFAULT_USER_GROUP], ro=[])
+
+        def _default_volume(res):
+            x = str(host_home_d.join("{username}"))
+            if x not in res:
+                res[x] = PKDict(
+                    bind=str(guest_home_d.join("jupyter")),
+                    mode=_default_mode(),
+                )
+            return res
+
+        def _normalize_and_default():
+            res = PKDict()
+            for n, v in volumes.items():
+                try:
+                    if not isinstance(v, dict):
+                        x = PKDict(bind=v)
+                    else:
+                        assert "bind" in v, f"bind must be specified in volume={n}: {v}"
+                        x = copy.deepcopy(v)
+                        if "mode" in x:
+                            x.mode.pksetdefault(rw=list)
+                            x.mode.pksetdefault(ro=list)
+                    x.pksetdefault(mode=_default_mode)
+                    res[n] = x
+                except Exception:
+                    pkdlog("volume={} config={}", n, v)
+                    raise
+            return _default_volume(res)
+
+        def _users(groups, host_path):
+            if not groups:
+                return set()
+            return _users_for_groups(groups, "volume", host_path, user_groups)
+
+        self._cfg = _normalize_and_default()
+        self._list = _cfg_to_list()
+
+    def for_user_sorted_by_guest_path(self, user):
+        """Find volumes matching user
+
+        Args:
+            user (str): matches jupyterhub name
+        Returns:
+            list: Volume instances in guest path sorted order
+        """
+
+        def _fmt(path):
+            return str(path).format(username=user)
+
+        res = PKDict()
+        for v in self._list:
+            if v.guest in res:
+                # These loops are in precedence order
+                continue
+            # user overrides _EVERYBODY
+            for n in user, _EVERYBODY:
+                # rw overrides ro
```

Comment:
You are certainly correct, because I really was confused about the code. It took me quite a while to figure out how to [code this](https://github.com/radiasoft/rsconf/pull/339/files#diff-133f161cbecd221f964ed3b4f3e9e26421d6018bb9e086b4214d1071c7667781R304). Better?

620d5dff